x-pack/plugins/enterprise_search/public/applications/app_search/components/document_creation/document_creation_buttons.tsx

```diff
@@ -0,0 +1,129 @@
+
+interface Props {
+  disabled?: boolean;
+}
+
+export const DocumentCreationButtons: React.FC<Props> = ({ disabled = false }) => {
+  const { openDocumentCreation } = useActions(DocumentCreationLogic);
+
+  const { engineName, isSampleEngine } = useValues(EngineLogic);
+  const {
+    myRole: { canViewEngineCrawler },
+  } = useValues(AppLogic);
+  const showCrawlerLink = canViewEngineCrawler && !isSampleEngine;
+  const crawlerLink = getEngineRoute(engineName) + ENGINE_CRAWLER_PATH;
+
+  return (
+    <>
+      <EuiText color="subdued">
+        <p>
+          <FormattedMessage
+            id="xpack.enterpriseSearch.appSearch.documentCreation.description"
+            defaultMessage="There are four ways to send documents to your engine for indexing. You can paste raw JSON, upload a {jsonCode} file, {postCode} to the {documentsApiLink} endpoint, or test drive the new Elastic Crawler (beta) to automatically index documents from a URL. Click on your choice below."
+            values={{
+              jsonCode: <EuiCode>.json</EuiCode>,
+              postCode: <EuiCode>POST</EuiCode>,
+              documentsApiLink: (
+                <EuiLink target="_blank" href={`${DOCS_PREFIX}/indexing-documents-guide.html`}>
+                  documents API
+                </EuiLink>
+              ),
+            }}
+          />
+        </p>
+      </EuiText>
+      <EuiSpacer />
+      <EuiFlexGrid columns={showCrawlerLink ? 2 : 3}>
+        <EuiFlexItem>
+          <EuiCard
+            title={i18n.translate(
+              'xpack.enterpriseSearch.appSearch.documentCreation.buttons.text',
+              { defaultMessage: 'Paste JSON' }
+            )}
+            description=""
+            icon={<EuiIcon type="indexEdit" size="xxl" color="primary" />}
+            data-test-subj="IndexingPasteJSONButton"
+            onClick={() => openDocumentCreation('text')}
+            isDisabled={disabled}
+          />
+        </EuiFlexItem>
+        <EuiFlexItem>
+          <EuiCard
+            title={i18n.translate(
+              'xpack.enterpriseSearch.appSearch.documentCreation.buttons.file',
+              { defaultMessage: 'Upload a JSON file' }
+            )}
+            description=""
+            icon={<EuiIcon type="exportAction" size="xxl" color="primary" />}
+            onClick={() => openDocumentCreation('file')}
+            isDisabled={disabled}
+          />
+        </EuiFlexItem>
+        <EuiFlexItem>
+          <EuiCard
+            title={i18n.translate('xpack.enterpriseSearch.appSearch.documentCreation.buttons.api', {
+              defaultMessage: 'Index from API',
+            })}
+            description=""
+            icon={<EuiIcon type="editorCodeBlock" size="xxl" color="primary" />}
+            onClick={() => openDocumentCreation('api')}
+            isDisabled={disabled}
+          />
+        </EuiFlexItem>
+        {showCrawlerLink && (
+          <EuiFlexItem>
+            <EuiCardTo
```

Comment:
There's so many different ways that we can link to stuff. It confuses me quite a bit. 

```
import { useHistory } from 'react-router-dom';
...
const history = useHistory();
...
<EuiCard onClick={() => history.push(crawlerLink)}
```

```
import { KibanaLogic } from './kibana_logic';
...
const { navigateToUrl } = useValues(KibanaLogic);
...
<EuiCard onClick={() => navigateToUrl(crawlerLink)}
```

```
import { KibanaLogic } from './kibana_logic';
...
const { history } = useValues(KibanaLogic);
...
<EuiCard onClick={() => history.push(crawlerLink)}
```

```
import { ReactRouterHelper } from 'wherever';
...
<ReactRouterHelper to={crawlerLink}>
  <EuiCard onClick={() => history.push('crawlerLink')}
</ReactRouterHelper>
```

You chose to create a component to encapsulate the last option there. What is the significance of that approach?

Here's some information that I dug up:

[Here](https://www.elastic.co/guide/en/kibana/7.10/kibana-navigation.html#navigating-between-kibana-apps) it seems to suggest using `navigateToUrl` is preferable to setting `window.location` because it is SPA friendly.

However, using `history.push` is also SPA friendly, right? It's even mentioned in [this code](https://github.com/elastic/eui/blob/master/wiki/react-router.md#react-router-51) that you linked to, Constance, and I've seen used in multiple places throughout Kibana, even in some of our own Enterprise Search code.

However, the [docs](https://github.com/elastic/kibana/blob/7.10/docs/development/core/public/kibana-plugin-core-public.applicationstart.navigatetourl.md) for `navigateToUrl` mention some significance of `basePath` in the equation.

So my assumption is that you use `navigateToUrl` instead of `history.push` because there may or may not be a base path (like if you're using spaces) in your URL. `navigateToUrl` will account for that base path in some way when using absolute paths?

The second reason for that assumption is the new "Running Kibana in dev mode with --no-base-path disables several useful features and is not recommended" message at Kibana startup, and a subsequent answer I saw someone post on Slack:

> Question: what are these “several useful features” that aren’t available without a base path
> Answer: The main problem with running Kibana without a base-path that people forget to use it in their code that leads to broken links in the runtime.

It makes me think, as I've seen places in our own code base where we use `history.push`, if our code is going to 
break in. places once it's run out of dev mode.

It also just kind of shocks me how much code we need just to navigate from one page to another. Like the whole concept of `ReactRouterHelper` is a little eyebrow raising. I also don't know why we use that instead of https://github.com/elastic/kibana/blob/7.10/src/plugins/kibana_react/public/app_links/redirect_app_link.tsx#L49 which appears to do the same thing.

Sorry for the long, meandering thought stream, I'm just trying to wrap my head around this whole thing.

Reply:
Talked to @constancecchen offline.

- Using the global `useHistory` and `history.push` will force a full re-render since Kibana stores state.
- Using Kibana's `history` directly could be feasible but it's still better to use `navigatToUrl`.
- Constance's helpers just make it a whole lot easier to implement these links, it does a bunch of calculations to maintain the base path and is much more declarative.
- EUI docs can sometimes be confusing since EUI is agnostic to both react-router and Kibana.